include/envoy/common/platform.h

```diff
@@ -288,3 +288,16 @@ struct mmsghdr {
 // On non-Linux platforms use 128 which is libevent listener default
 #define ENVOY_TCP_BACKLOG_SIZE 128
 #endif
+
+// Used for retrieving TCP round-trip time.
+#if defined(__linux__)
+// On Linux, default to the actual TCP_INFO defined in kernel's netinet/tcp.h header.
+#define ENVOY_TCP_INFO TCP_INFO
+#else
+// Define ENVOY_TCP_INFO to an unused value.
+#define ENVOY_TCP_INFO -1
+// Define tcp_info to a struct with just the fields needed to compile on other platforms.
+struct tcp_info {
+  uint32_t tcpi_rtt;
+};
```

Comment:
You probably already had this conversion (sorry) but it seems strange/fragile to fake this, just to force an error return from the socket API. Would this be better handled by checking if the capability is defined and doing the socket call, otherwise just returning nullopt?

Reply:
Do you mean changing the implementation in `source/common/network/listen_socket_impl.h` to something like:

```
  absl::optional<std::chrono::milliseconds> lastRoundTripTime() override {
    #if defined TCP_INFO
    struct tcp_info ti;
    socklen_t len = sizeof(ti);
    if (!SOCKET_FAILURE(ioHandle().getOption(IPPROTO_TCP, TCP_INFO, &ti, &len).rc_)) {
      return std::chrono::milliseconds(ti.tcpi_rtt);
    }
    #endif

    return {};
  }
```

Either is fine with me, I don't think I have a good grasp of why we might prefer one over the other.

Reply:
Yes that's what I meant. Personally I think the latter is less fragile vs. forcing a socket call that you expect to fail just to return fake data, but I don't feel strongly about it. @florincoras @yanavlasov any major opinion here?

Reply:
The goal, at least in my mind, was that maybe somebody will at one point catch "TCP_INFO" getsockopt request on windows/macos, do the right syscall(s) and eventually return something useful in tcpi_rtt. I guess we could still do that now and return an error without doing the syscall. Still, the above seems simple and clean enough to me as we avoid per platform versions of lastRoundTripTime.

Reply:
Can we somehow avoid forcing the failing syscall on platforms that don't support it? That's the part that seems fragile. Should perhaps this should be a direct ioHandle() API vs. using getOption() and then it can easily return nothing on non-linux?

Reply:
If we do something like `if (optname == TCP_INFO) { return {-1, ENOTSUP/ENOSYS}; }` before calling `::getsockopt` in `os_sys_calls_impl.cc` we should be able to avoid the syscall, for windows at least. Then, when somebody figures out how to support it, it could be changed to return the actual value. We can also go with per `IoHandle` implementations if we want to avoid the extra`if` but it will be a more involved change. Any of the three options (the one proposed by @fcfort above being the first) work for me.

Reply:
My personal preference is to just make this an explicit part of the IoHandle interface, and then we can easily override it on a per-platform basis?

Reply:
To be clear, the proposal is to add `lastRoundTripTime` to the `IoHandle` interface?

Reply:
This is what I would suggest. Even better, define a cross platform `IoHandleInfo` struct, with a single optional field, `last_round_trip_time`. Populate this field on Linux/TCP and don't populate it everywhere else. This would be quite future proof.

Reply:
Hmm, why do we need to make `last_round_trip_time` part of a structure `IoHandleInfo`? In case some other members are added to the `tcp_info` structure?

Reply:
Yes, just to future proof it. But not a big deal. Up to you.

Reply:
I took a stab at it, let me know if that's what you're expecting. I decided to skip creating a new `IoHandleInfo` structure.

Reply:
I think this entire change in this file can be reverted now.

Reply:
Done, reverted the changes to `platform.h`.